File: gymflow-frontend/src/services/memberService.js
import apiClient from '../apiClient';
import { MEMBER_ENDPOINTS } from '../apiConfig';

export const getMembers = async () => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_ALL);
    return response.data;
  } catch (error) {
    console.error("Error fetching members", error);
    throw error;
  }
};

export const getMemberById = async (id) => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_BY_ID(id));
    return response.data;
  } catch (error) {
    console.error(`Error fetching member ${id}`, error);
    throw error;
  }
};

export const getMemberByEmail = async (email) => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_BY_EMAIL, { params: { email } });
    return response.data;
  } catch (error) {
    console.error(`Error fetching member by email ${email}`, error);
    throw error;
  }
};

export const createMember = async (memberData) => {
    try {
        const response = await apiClient.post(MEMBER_ENDPOINTS.CREATE, memberData);
        return response.data;
    } catch (error) {
        console.error("Error creating member", error);
        throw error;
    }
};

export const updateMember = async (id, memberData) => {
  try {
    const response = await apiClient.put(MEMBER_ENDPOINTS.UPDATE(id), memberData);
    return response.data;
  } catch (error) {
    console.error(`Error updating member ${id}`, error);
    throw error;
  }
};

export const deleteMember = async (id) => {
  try {
    const response = await apiClient.delete(MEMBER_ENDPOINTS.DELETE(id));
    return response.data;
  } catch (error) {
    console.error(`Error deleting member ${id}`, error);
    throw error;
  }
};

export const activateMember = async (id) => {
  try {
    const response = await apiClient.put(MEMBER_ENDPOINTS.ACTIVATE(id));
    return response.data;
  } catch (error) {
    console.error(`Error activating member ${id}`, error);
    throw error;
  }
};

export const deactivateMember = async (id) => {
  try {
    const response = await apiClient.put(MEMBER_ENDPOINTS.DEACTIVATE(id));
    return response.data;
  } catch (error) {
    console.error(`Error deactivating member ${id}`, error);
    throw error;
  }
};

export const logAttendance = async (memberId, recordedBy = 'RECEPTIONIST') => {
  try {
    const response = await apiClient.post(MEMBER_ENDPOINTS.LOG_ATTENDANCE(memberId), { recordedBy });
    return response.data;
  } catch (error) {
    console.error(`Error logging attendance for member ${memberId}`, error);
    throw error;
  }
};

export const renewMembership = async (id, newStartDate) => {
  try {
    const response = await apiClient.post(MEMBER_ENDPOINTS.RENEW_MEMBERSHIP(id), null, { params: { newStartDate } });
    return response.data;
  } catch (error) {
    console.error(`Error renewing membership for member ${id}`, error);
    throw error;
  }
};

export const getMembersByStatus = async (status) => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_BY_STATUS(status));
    return response.data;
  } catch (error) {
    console.error(`Error fetching members by status ${status}`, error);
    throw error;
  }
};

export const getMembersByMembershipStatus = async (status) => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_BY_MEMBERSHIP_STATUS(status));
    return response.data;
  } catch (error) {
    console.error(`Error fetching members by membership status ${status}`, error);
    throw error;
  }
};

export const getMembersEndingBefore = async (date) => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_ENDING_BEFORE, { params: { date } });
    return response.data;
  } catch (error) {
    console.error('Error fetching members ending before', error);
    throw error;
  }
};

export const getMembersEndingBetween = async (start, end) => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_ENDING_BETWEEN, { params: { start, end } });
    return response.data;
  } catch (error) {
    console.error('Error fetching members ending between', error);
    throw error;
  }
};

export const getMembersWithOutstandingPayments = async () => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_OUTSTANDING_PAYMENTS);
    return response.data;
  } catch (error) {
    console.error('Error fetching members with outstanding payments', error);
    throw error;
  }
};

export const getMembersWithNoPayments = async () => {
  try {
    const response = await apiClient.get(MEMBER_ENDPOINTS.GET_NO_PAYMENTS);
    return response.data;
  } catch (error) {
    console.error('Error fetching members with no payments', error);
    throw error;
  }
};

File: gymflow/src/main/java/com/application/gymflow/service/MemberService.java
package com.application.gymflow.service;

import com.application.gymflow.dto.member.MemberRequestDto;
import com.application.gymflow.dto.member.MemberUpdateRequestDto;
import com.application.gymflow.enums.common.Status;
import com.application.gymflow.enums.member.AttendanceStatus;
import com.application.gymflow.enums.member.MembershipStatus;
import com.application.gymflow.enums.member.RecordedByType;
import com.application.gymflow.model.member.AttendanceLog;
import com.application.gymflow.model.member.Member;

import java.time.LocalDate;
import java.util.List;

/**
 * Service interface for managing Member entities.
 *
 * <p>This interface defines the core business logic for member-related operations,
 * including creation, retrieval, updates, deletion, and management of
 * associated entities like memberships, payments, and attendance logs.
 * </p>
 */
public interface MemberService {

    /**
     * Creates a new member and all associated records (user, membership, fitness profile, payment).
     *
     * @param requestDto The DTO containing the member's details.
     * @return The newly created Member entity.
     */
    Member createMember(MemberRequestDto requestDto);

    /**
     * Retrieves a member by their unique ID.
     *
     * @param id The ID of the member.
     * @return The found Member entity.
     */
    Member getMemberById(Long id);

    /**
     * Retrieves a list of all members in the system.
     *
     * @return A list of Member entities.
     */
    List<Member> getAllMembers();

    /**
     * Updates an existing member's details.
     *
     * @param id The ID of the member to update.
     * @param updateDto The DTO containing the updated details.
     * @return The updated Member entity.
     */
    Member updateMember(Long id, MemberUpdateRequestDto updateDto);

    /**
     * Deletes a member by their ID.
     *
     * @param id The ID of the member to delete.
     */
    void deleteMember(Long id);

    /**
     * Retrieves a member by their email address.
     *
     * @param email The email address of the member.
     * @return The found Member entity.
     */
    Member getMemberByEmail(String email);

    /**
     * Logs the attendance of a member.
     *
     * @param memberId The ID of the member.
     * @param recordedBy The type of entity that recorded the attendance (e.g., RECEPTIONIST).
     * @return The created AttendanceLog entity.
     */
    AttendanceLog logAttendance(Long memberId, RecordedByType recordedBy);

    /**
     * Retrieves a list of members whose membership ends before a given date.
     *
     * @param date The date to check against.
     * @return A list of Member entities.
     */
    List<Member> getMembersWithMembershipEndingBefore(LocalDate date);

    /**
     * Retrieves a list of members whose membership ends within a date range.
     *
     * @param start The start date of the range (inclusive).
     * @param end The end date of the range (inclusive).
     * @return A list of Member entities.
     */
    List<Member> getMembersWithMembershipEndingBetween(LocalDate start, LocalDate end);

    /**
     * Retrieves a list of members who have not made any payments.
     *
     * @return A list of Member entities.
     */
    List<Member> getMembersWithNoPayments();

    /**
     * Activates a member and their associated membership.
     *
     * @param memberId The ID of the member to activate.
     * @return The activated Member entity.
     */
    Member activateMember(Long memberId);

    /**
     * Deactivates a member and their associated membership.
     *
     * @param memberId The ID of the member to deactivate.
     * @return The deactivated Member entity.
     */
    Member deactivateMember(Long memberId);

    /**
     * Renews a member's membership, setting a new start and end date.
     *
     * @param memberId The ID of the member.
     * @param newStartDate The new start date for the renewed membership.
     * @return The Member entity with the renewed membership.
     */
    Member renewMembership(Long memberId, LocalDate newStartDate);

    /**
     * Retrieves a list of members with outstanding payments.
     *
     * @return A list of Member entities.
     */
    List<Member> getMembersWithOutstandingPayments();

    /**
     * Retrieves a list of members based on their membership status.
     *
     * @param status The membership status to filter by.
     * @return A list of Member entities.
     */
    List<Member> getMembersByMembershipStatus(MembershipStatus status);

    /**
     * Retrieves a list of members based on their overall status (ACTIVE/INACTIVE).
     *
     * @param status The member status to filter by.
     * @return A list of Member entities.
     */
    List<Member> getMembersByStatus(Status status);
}
